CMSgov/easi-app
# EASI-2513 - Add TRB Admin Notes

# EASI-2513

## Changes and Description

- Introduce TRB Admin Notes
    - Add SQL migration to create table for them
    - Create store methods for creating/fetching/updating them
        - These methods all follow the conventions we've established - they take a `context.Context` instead of just a logger, `GetTRBAdminNoteByID()` returns `nil, nil` if it doesn't find a matching note instead of logging an error (the resolver returns an error instead).
    - Add them to GraphQL schema
    - Add resolvers to link GQL operations with store methods


## Notes
* ~~I still want to add tests for the store methods, but I can get this up for review and running in CI first.~~ Tests added in https://github.com/CMSgov/easi-app/pull/1879/commits/4a35d68b27e75ccf0a31a8561352ed66719ff60c.
* I included `author: UserInfo!` (which includes EUA ID, common name, and email) as a field in the GraphQL type for `TRBAdminNote`, so the frontend doesn't have to make another request to look up the author's name based on their EUA ID.
* For handling archived (soft-deleted) notes, after a [discussion in Slack](https://cmsgov.slack.com/archives/CNU2B59UH/p1673376790722429), I decided to always return notes regardless of archived status for right now. Once the UX and requirements for archived notes are more developed and fully-formed (post-MVP), if we want/need to implement filtering on the backend, we can. The comments on the Jira ticket have some ideas for how to design this.
* Similar to https://github.com/CMSgov/easi-app/pull/1865 and https://github.com/CMSgov/easi-app/pull/1866, I didn't add any unit tests for the resolver functions due to their simplicity.
* ~~The tests added a utility function `createTRBRequestWithContext()`; if https://github.com/CMSgov/easi-app/pull/1877 is merged before this PR, that'll become redundant. We can move the existing `createTRBRequest()` function from `pkg/storage/trb_advice_letter_test.go` into `pkg/storage/store_test.go`, and use that function for this PR's tests as well.~~ Done as of https://github.com/CMSgov/easi-app/pull/1879/commits/db714aef395683c3ce33fb04d03951ef9e77b5fe.


## Questions for Reviewers
* Naming conventions - unlike most of the TRB-related models, this just uses a prefix of "TRB" instead of "TRBRequest" - it's `TRBAdminNote`/`trb_admin_note` instead of `TRBequestAdminNote`/`trb_request_admin_note`. If we want to change it to have the "Request" part of the prefix for consistency, we can. If we do that, we'll probably want to do the same for `TRBAdviceLetter`/`trb_advice_letter` instead of dropping the prefix.


## How to test this change

1. Spin up the backend and open the GraphQL playground.
2. Create a TRB request, noting the ID for later:
```gql
mutation {
  createTRBRequest(requestType: NEED_HELP) {
    id
    type
  }
}
```
3. Query for the notes on the request - there should be no notes (since none have been created), but no errors. (Save this query, it'll be used again)
```gql
query {
  trbRequest(id:"trbRequestIdFromAbove") {
    adminNotes {
      id
      trbRequestId
      isArchived
      category
      noteText
      author {
        euaUserId
        commonName
        email
      }
      createdBy
      createdAt
      modifiedBy
      modifiedAt
    }
  }
}
```
4. Create an admin note.
```gql
mutation createFirstNote {
  createTRBAdminNote(
    input: {
      trbRequestId: "trbRequestIdFromAbove"
      category: GENERAL_REQUEST
      noteText: "First note - Needs some more polish"
    }
  ) {
    id
    trbRequestId
    isArchived
    category
    noteText
    author {
      euaUserId
      commonName
      email
    }
    createdBy
    createdAt
    modifiedBy
    modifiedAt
  }
}
```
5. Run the query from step 3 again. You should see the created note, with the correct category and text, and `isArchived` set to `false`.
6. Create a second admin note.
```gql
mutation createSecondNote {
  createTRBAdminNote(
    input: {
      trbRequestId: "trbRequestIdFromAbove"
      category: ADVICE_LETTER
      noteText: "Second note - Maybe soften the criticism a bit"
    }
  ) {
    id
    trbRequestId
    isArchived
    category
    noteText
    author {
      euaUserId
      commonName
      email
    }
    createdBy
    createdAt
    modifiedBy
    modifiedAt
  }
}
```
7. Run the query from step 3 again. You should see both created notes, with distinct data for each note.
8.  Update the first admin note, changing the local authentication header in the GraphQL playground to use `TEST` as the EUA ID, using the note ID returned from creating the note in step 4.
```gql
mutation updateFirstNote {
  updateTRBAdminNote(
    input: {
      id: "trbAdminNoteFromStep4"
      category: INITIAL_REQUEST_FORM
      noteText: "First note - updated"
    }
  ) {
    id
    trbRequestId
    isArchived
    category
    noteText
    author {
      euaUserId
      commonName
      email
    }
    createdBy
    createdAt
    modifiedBy
    modifiedAt
  }
}
```
9. Run the query from step 3 again. The first note should show the updated values for the `category` and `noteText` fields, the `modifiedBy` field should be `"TEST"`, and `modifiedAt` should show the time when the update mutation was ran.
10. Archive the second admin note, changing the local authentication header in the GraphQL playground to use `GRTB` as the EUA ID, using the ID returned from creating the note in step 6.
```gql
mutation archiveSecondNote {
  archiveTRBAdminNote(id: "trbAdminNoteIdFromStep6") {
    id
    trbRequestId
    isArchived
    category
    noteText
    author {
      euaUserId
      commonName
      email
    }
    createdBy
    createdAt
    modifiedBy
    modifiedAt
  }
}

```
11. Run the query from step 3 again. The second note should show that `isArchived` is now `true`, the `modifiedBy` field should be `"GRTB"`, and `modifiedAt` should show the time when the archival mutation was ran.
